src/beast/beast/http/basic_parser.h

```diff
@@ -113,153 +133,405 @@ class basic_parser
+        @param size The number of bytes in the buffer pointed to by data.
+        @param ec Set to the error, if any error occurred.
+
+        @return The number of bytes consumed in the input sequence.
+    */
+    std::size_t
+    write(void const* data, std::size_t size,
+        error_code& ec);
+
+    /** Write data to the parser.
+
+        @param buffers An object meeting the requirements of
+        ConstBufferSequence that represents the input sequence.
+
+        @throws boost::system::system_error Thrown on failure.
+
+        @return The number of bytes consumed in the input sequence.
+    */
+    template<class ConstBufferSequence>
+    std::size_t
+    write(ConstBufferSequence const& buffers)
+    {
+        error_code ec;
+        auto const used = write(buffers, ec);
+        if(ec)
+            throw boost::system::system_error{ec};
+        return used;
+    }
+
+    /** Write data to the parser.
+
+        @param buffers An object meeting the requirements of
+        ConstBufferSequence that represents the input sequence.
+        @param ec Set to the error, if any error occurred.
+
+        @return The number of bytes consumed in the input sequence.
     */
-    template <class ConstBufferSequence>
-    std::pair <error_code, std::size_t>
-    write (ConstBufferSequence const& buffers);
+    template<class ConstBufferSequence>
+    std::size_t
+    write(ConstBufferSequence const& buffers,
+        error_code& ec);
 
     /** Called to indicate the end of file.
+        
         HTTP needs to know where the end of the stream is. For example,
         sometimes servers send responses without Content-Length and
         expect the client to consume input (for the body) until EOF.
         Callbacks and errors will still be processed as usual.
+        
         @note This is typically called when a socket read returns eof.
-        @return `true` if the message is complete.
+        
+        @throws boost::system::system_error Thrown on failure.
     */
-    error_code
-    write_eof();
+    void
+    write_eof()
+    {
+        error_code ec;
+        write_eof(ec);
+        if(ec)
+            throw boost::system::system_error{ec};
+    }
 
-protected:
-    /** Called once when a new message begins. */
-    virtual
+    /** Called to indicate the end of file.
+        
+        HTTP needs to know where the end of the stream is. For example,
+        sometimes servers send responses without Content-Length and
+        expect the client to consume input (for the body) until EOF.
+        Callbacks and errors will still be processed as usual.
+        
+        @note This is typically called when a socket read returns eof.
+        
+        @param ec Set to the error, if any error occurred.
+    */
     void
-    on_start() = 0;
+    write_eof(error_code& ec);
+
+private:
+    Derived&
+    impl()
+    {
+        return *static_cast<Derived*>(this);
```

Comment:
I took a look at `parser` and I don't see where CRTP is needed for that class. We could get the same behavior by templating `basic_parser` with `parser` and keeping a member var without ever having `parser` inherit from `basic_parser`. Did I miss where it's needed in `parser` or are there other classes that need the full power of CRTP?